Refactor some controlled component stuff

sebmarkbage opened the pull request:
This is mainly renaming some stuff. The behavior change is hasOwnProperty to nullish check.

I had a bigger refactor that was a dead-end but might as well land this part and see if I can pick it up later.

sebmarkbage commented on packages/react-dom-bindings/src/client/ReactDOMComponent.js:
```diff
@@ -2263,38 +2270,47 @@ export function diffHydratedProperties(
       // post mount to force it to diverge from attributes. However, for
       // option and select we don't quite do the same thing and select
       // is not resilient to the DOM state changing so we don't do that here.
       // TODO: Consider not doing this for input and textarea.
```
There's something off about these inconsistencies. We probably should be issuing change events if the value has changed before hydration and then reset using the "restore" mechanism. But the fact that select is different is also weird. Does that mean that controlled selects are not updated to reflect the "state" if it changes before hydration. Seems like it.

If we ironed this out we could decide whether the init should be merged with the postInit.

sebmarkbage replied:
I think maybe the other bug here is that we call postInit during hydration, and not in the commit phase in Fiber which seems wrong too. If it was in the commit phase the "mount" name would make more sense.

sophiebits reviewed (approved):
Is there a bug in the existing code when removing defaultValue? This fails for me:

```js
  it('should remove previous `defaultValue`', () => {
    const node = ReactDOM.render(
      <input type="text" defaultValue="0" />,
      container,
    );

    expect(node.value).toBe('0');
    expect(node.defaultValue).toBe('0');

    ReactDOM.render(<input type="text" />, container);
    expect(node.defaultValue).toBe('');
  });
```

```
 FAIL  packages/react-dom/src/__tests__/ReactDOMInput-test.js (7.948 s)
  ● ReactDOMInput › should remove previous `defaultValue`

    Expected: ""
    Received: "0"

      461 |
      462 |     ReactDOM.render(<input type="text" />, container);
    > 463 |     expect(node.defaultValue).toBe('');
          |                               ^
      464 |   });
      465 |
```

sebmarkbage commented:
Tbh, I feel like there are a lot of subtle semantics that are wrong here. However, for that particular scenario I could see how that would be expected given the "attribute syncing" since defaultValue is the same as the attribute, it means that since the internal state of the value is still the stateful value, and if that is sync:ed back to the attribute then the defaultValue becomes that too.

If gated on the disableInputAttributeSyncing it does seems like that should work though.

sophiebits commented:
> since defaultValue is the same as the attribute, it means that since the internal state of the value is still the stateful value, and if that is sync:ed back to the attribute then the defaultValue becomes that too

Not sure I understand what you mean here. FWIW updating the defaultValue prop does set the attribute regardless of the feature flag so I expected that removing the prop would behave the same way:

https://github.com/facebook/react/blob/657698e48d5b093b4ea5a7684e26d3fdd16695f2/packages/react-dom/src/__tests__/ReactDOMInput-test.js#L449

sebmarkbage commented:
I see, yea, if it has never transitioned to having a controlled or stateful "value" then it should.

sebmarkbage commented:
It seems like the implementation is not consistent about `hasOwnProperty`, `undefined` and `null`. For example, we treat `null` as something that should fall back to the `defaultValue` if it's available, but in other cases, we treat only `undefined` as missing. This leads to a combination of cases that only happen with certain types of transitions. However it mostly comes with a warning if you specify `null`.

sebmarkbage commented:
Ok, I did some further refactoring. I removed all the `_wrapperState` to show that they're not actually used to store any meaningful state. They're just derived from initial props. Since defaultValue does update.

I also fixed it so that removing or switching defaultValue to null or empty does update on input and textarea. Since textarea doesn't have an attribute but the child content, I just update to an empty string for this case.

sebmarkbage commented on packages/react-dom-bindings/src/client/ReactDOMTextarea.js:
```diff
@@ -57,20 +49,48 @@ export function initWrapperState(element: Element, props: Object) {
+    if (newValue !== node.value) {
+      node.value = newValue;
+    }
+    // TOOO: This should respect disableInputAttributeSyncing flag.
```
@gaearon This doesn't update an attribute but it does update the internal state so it falls under the same category of issues.

sebmarkbage commented on packages/react-dom/src/__tests__/ReactDOMInput-test.js:
```diff
@@ -1952,11 +1952,7 @@ describe('ReactDOMInput', () => {
-      } else {
-        expect(input.getAttribute('value')).toBe('first');
-      }
+      expect(input.getAttribute('value')).toBe(null);
```
This is considered a bug fix.

sophiebits commented on packages/react-dom-bindings/src/client/ReactDOMInput.js:
```diff
@@ -302,13 +308,20 @@ export function postInitInput(
     node.name = '';
   }
 
+  const defaultChecked =
+    props.checked != null ? props.checked : props.defaultChecked;
```
seems a little confusing that your defaultValue and defaultChecked are not parallel (the latter here incorporates props.checked) but I guess it's fine

sophiebits commented on packages/react-dom-bindings/src/events/plugins/ChangeEventPlugin.js:
```diff
@@ -335,8 +336,12 @@ function extractEvents(
+    // These props aren't necessarily the most current but we warn for changing
+    // between controlled and uncontrolled, so it doesn't matter and the previous
+    // code was also broken for changes.
+    const props = targetInst.memoizedProps;
```
nit: pass isControlled (props.value != null) here instead of having it in the callee? since the props can be stale, seems better to localize the calculations to the same place where we know they are stale

sophiebits commented on packages/react-dom-bindings/src/client/ReactDOMInput.js:
```diff
@@ -93,32 +79,18 @@ export function initInput(element: Element, props: Object) {
-  const defaultValue = props.defaultValue == null ? '' : props.defaultValue;
-  const initialChecked =
-    props.checked != null ? props.checked : props.defaultChecked;
-  node._wrapperState = {
```
nice

sophiebits reviewed (commented):
lgtm
